### ISSUE TITLE
[BUG - Experiemental] RegisterHook generally not working in Grounded

### ISSUE THREAD

**Issue by Aidanamite** [UE4SS-RE/RE-UE4SS#626]
**Branch or Release**
Experimental 3.0.1-151

**Game and Engine Version**
Grounded
UE 4.27

**Describe the bug**
Most of the functions in Grounded do not seem to be able to be hooked. Some can, but most not. Calling the RegisterHook function says that it was successful in the log, but when the function is supposed to occur, the function provided to the hook is simply not run.

I have attempted to fix the issue by both configuring the engine version override to the game's engine version and delaying registering the hooks even as late as after I've loaded into a world.

Based on what I can tell from using the zDev tools for UE4SS, the functions that are not hookable are any function that is both `Final` and `BlueprintCallable`

Using the UE4SS header dumps, I was able to use Cheat Engine to find the function and add a breakpoint to verify that the function is being called and it is, just that the hook itself isn't

**To Reproduce**
1. Use the RegisterHook function on any function that is Final & BlueprintCallable
2. Do something that would trigger that function

In my case, the main function I've been trying to hook is the `/Script/Maine.Item:RemoveDurability` function

**Expected behavior**
I expected the hook function to be run.

**Desktop (please complete the following information):**
 - OS: Win10

**Comment by iMagic16** [comment]
+1 UE4SS is not working in Grounded 1.4.5+

**Comment by jgold189** [comment]
@Aidanamite can you possibly explain your steps a bit more in-depth for how you confirmed the function were still being called even though the registered hook was never triggering? I'm trying to figure out if I'm running into the same issue when trying to mod Echo Point Nova which is unreal engine version 4.27.2.0. I'm on the 3.0.1 release so slightly different from the experimental branch but I checked 2.5.2 and nothing was different.

**Comment by Aidanamite** [comment]
> can you possibly explain your steps a bit more in-depth for how you confirmed the function were still being called even though the registered hook was never triggering?

i used cheat engine to manually locate the function's native code and used breakpoints to check if the code worked how i thought it did

**Comment by Aidanamite** [comment]
as a side note to that: i've been "modding" the game using cheat engine for a while now, making fairly complex scripts to add in various behaviours.

i was hoping to be able to use ue4ss to remake the same sort of tweaks in a more dynamic and maintainable way

**Comment by igromanru** [comment]
> as a side note to that: i've been "modding" the game using cheat engine for a while now, making fairly complex scripts to add in various behaviours.
> 
> i was hoping to be able to use ue4ss to remake the same sort of tweaks in a more dynamic and maintainable way

Make a UE4SS C++ mod then. It has more freedom, Lua implementation is still very limited.

**Comment by Aidanamite** [comment]
can the c++ mods hook functions? as far as i can tell, the issue is that it's failing to hook the functions, therefore wouldn't it have the same issue regardless of if i'm using lua or c++ or does it use a completely different hooking system for c++ mods?

**Comment by igromanru** [comment]
> can the c++ mods hook functions that lua can't? as far as i can tell, the issue is that it's failing to hook the functions, therefore wouldn't it have the same issue regardless of if i'm using lua or c++ or does it use a completely different hooking system for c++ mods?

Lua is just basically an API to call C++ functions, but it's not just a simple 1:1 call, it's more complicated and depends on the implementation. 
UE4SS C++ has much more options that you can do. In worst case you can always use `RegisterProcessEventPreCallback` or `RegisterProcessEventPostCallback` and catch the call of your function this way.  
You can even hook native or vtable functions that are not reflected, but it would obviously require some reverse engineering.  
Long story short, in C++ you can do anything.

**Comment by Aidanamite** [comment]
so do you think the issue of the lua hook not working is caused by the hook just failing to call the lua callback and not because the hook isn't attaching to the function correctly?

**Comment by igromanru** [comment]
> so do you think the issue of the lua hook not working is caused by the hook just failing to call the lua callback and not because the hook isn't attaching to the function correctly?

No, it's definitely not it. I just say that in a C++ mod you have much more options.  

After reading the first post once again, I think that the problem is that the game never calls the function over ProcessEvent, but calls it over it's native function.  
It is quite common that there are functions that are exposed to reflection but are only called by C++ code, which then uses its underlined native function. That's most likely the reason why you can breakpoint the function with CE but the hook in Lua never gets triggered.  
In such case the only solution is only to hook the native function with a detour hook, which is only possible with a C++ DLL inside the game.

**Comment by Aidanamite** [comment]
i assumed the hooks were detour hooks by default. because that's the more reliable way to ensure the hook gets called.

**Comment by UE4SS** [comment]
This is either a UE4SS bug, or the game is calling the native C++ function directly.
I'm labeling it as a bug so that this can be checked at some point.

We hook UE functions in two different ways:
1. If it's native: The `UFunction->Func` pointer is modified to point to our own function (effectively a detour) which then processes callbacks for Lua mods that have hooks registered for that particular function.
2. If it's non-native: We detour the internal UE function that processes BP function calls (`ProcessLocalScriptFunction` or `ProcessInternal` depending on the engine version) which then processes the callbacks.

The reason why native C++ functions don't get caught in our native hook is because if a game calls a native C++ function directly, then the corresponding `UFunction` object isn't used for that call, the compiler just generates a `CALL` or `JMP` instruction instead.
As far as I know, native C++ functions only get caught by our hook if it's called by non-native code.

**Comment by Aidanamite** [comment]
> or the game is calling the native C++ function directly
based on the asm code i was using directly before using ue4ss, i'm pretty sure most of the code uses direct function calls

**Comment by UE4SS** [comment]
Detouring arbitrary C++ functions via Lua isn't something we're planning on adding.

**Comment by PowerGamer1** [comment]
This really needs to be solved somehow. In Gothic 1 Remake a lot of reflected UE functions (from CXXHeaderDump) do not fire successfully installed LUA hooks. Even the hook from documented example does not work: https://docs.ue4ss.com/dev/lua-api/global-functions/registerkeybind.html.

**Comment by UE4SS** [comment]
> This really needs to be solved somehow. In Gothic 1 Remake a lot of reflected UE functions (from CXXHeaderDump) do not fire successfully installed LUA hooks. Even the hook from documented example does not work: https://docs.ue4ss.com/dev/lua-api/global-functions/registerkeybind.html.

It was already explained why this happens, I don't think there's anything to fix.
If you believe otherwise, find a function that's hookable in another game but not in Gothic 1 Remake (to prove that the function is supposed to work) and then make an issue specifically for that problem.
Also, you haven't defined what "fire successfully" means, so I've just assumed, because you replied to this issue, that the callback simply doesn't fire at all.
